accepted/2020/binary-data/binary-data.md

```diff
@@ -0,0 +1,195 @@
+BCL wouldn't be able to take `BinaryData`, but technologies above the BCL could
+(Microsoft.Extensions, ASP.NET Core, Azure SDK etc). While that might be
+unfortunate, we believe it's acceptable because most of the usability gains
+would be in higher levels anyway.
+
+Thus, we shouldn't think of this as an exchange type in the BCL. We already have
+exchange types; the point of this type is unify them by providing conversions;
+the point isn't to provide yet another type. Thus, we don't see the need to, for
+example, provide new virtual methods on Stream.
+
+We could decide to model the constructor that performs JSON serialization as a
+factory method. Combined with a hypothetical feature *extension statics* we
+could make the type more low level than serialization. An alternative design is
+to replace constructors and factories with instance members (e.g. `SetBytes`
+`SetObjectFromJson`) which would throw when the underlying array is non-null.
+That prevents mutation and uses a simple pattern
+
+```C#
+var data = new BinaryData();
+data.SetBytes(...);
+```
+
+This would also allow us to use regular extension methods to split out JSON
+serialization.
+
+We need to balance layering concerns with usability. The primary goal of this
+type is to provide usability via conveniences. If we don't make the type super
+usable, the feature has failed. While the design with extension methods feels
+quite appealing we need to verify that it realizes these goal. If we can't, we
+should go back to the design of using constructors and factory methods, even if
+that reduces our flexibility.
+
+### API design with extension methods
+
+```C#
+namespace System
+{
+    public readonly struct BinaryData
+    {
+        public void SetBytes(ReadOnlyMemory<byte> data);
+        public void SetBytes(ReadOnlySpan<byte> data);
+        public void SetBytes(byte[] data);
+        public void SetStream(Stream stream);
+        public static Task SetStreamAsync(Stream stream, CancellationToken cancellationToken = default);
+        public void SetString(string data);
+        public static implicit operator ReadOnlySpan<byte>(BinaryData data);
+        public static implicit operator ReadOnlyMemory<byte>(BinaryData data);
+        public ReadOnlyMemory<byte> ToBytes();
+        public Stream ToStream();
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override bool Equals(object? obj);
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override int GetHashCode();
+        public override string ToString();
+    }
+}
+namespace System.Text.Json
+{
+    public static class BinaryDataJsonExtensions
+    {
+        public static Task SetObjectAsJsonAsync<T>(this BinaryData data, T jsonSerializable, CancellationToken cancellationToken = default);
+        public static void SetObjectAsJson<T>(this BinaryData data, T jsonSerializable, CancellationToken cancellationToken = default);
```

Comment:
The first parameter should be annotated _ref this_, otherwise you'll end up mutating a copy of the struct, which is a glorified (and expensive) no-op. We may want to create an analyzer for people writing their own extension methods for this type.